### PR TITLE
#211 Improve mechanism for initialising database

### LINKED ISSUE #156: Add env variable for mock data
Mock data is currently loaded every time the server is started. This causes a duplicate primary key panic if the database already contains mock data. Would be good to add an env variable for this, e.g. `DATABASE_MOCK=true` (or something better).

### LINKED ISSUE #161: Insert mock transact data on startup
Mock `transact` data has been added, but is not currently being inserted into the database on startup.

### LINKED ISSUE #162: Insert mock transact_line data on startup
Mock `transact_line` data has been added, but is not currently being inserted into the database on startup.

### LINKED ISSUE #211: Improve mechanism for initialising database
Would be good to have a way to initialise the database with suitable mock data. 

Options:

- Insert mock data in code
- Run script to insert mock data via mutations
- Run script/`pg_dump` to directly initialise database

I think for now the last option works best? It's database dependent, but should be fine (unlikely we will be switching out DBs anytime soon!)

### PR DESCRIPTION
**Pull request by wlthomson**
Closes #211.
Closes #161.
Closes #162.
Closes #156.

Updates conditional compilation logic to enable running and testing the app using a mock database (implemented as a hashmap) and with mock data.

The app can be run using the mock database by enabling the `"mock`" feature, e.g.

```
cargo run --features mock
```

Tests can be run in the same way, e.g.

```
cargo test --features mock
```

Notice that `--features mock` can also be specified for tests. The idea is to allow for simplifying tests with a mocked database where appropriate, but leaving the door open for integration testing with a running database. This is a bit of a trade-off: it simplifies a lot of testing, but makes running tests a bit more complicated (have to run both `cargo test` and `cargo test --features mock` to get complete test coverage). Open to revisiting the idea if it turns out to be too complicated.

P.s. tried as much as possible to add descriptions to all my commits, should hopefully add some context to some of the changes!

### CODE REVIEW

**Review by andreievg — CHANGES_REQUESTED** [review]
Looks good, just a minor change request.

(haven't reviewed any mock stuff thought, would have been better in seperate PR imo). There is another way to do testing, which is fast and uses db, I'll make an issue with explanation. Anyways if we choose to go with mocks I'll give it a more thorough review in the future.

Also would be great if main method jumped up an abstraction level (still have logic in main.rs, but with helper methods), we can do this in seperate PR in the future, once we know what are all the bits that need to be in main

i.e. (something along these lines)
```
setupLogging();

let settings = getSetting();

let repository = getDBRepository(&settings)

startScheduler(&repository)

let syncConsumer = initiliseSync(&repository)

let startWebServer(&repository, &settings)
```

File: src/main.rs [hunk 1]
```diff
@@ -1,82 +1,191 @@
...
+    repositories.insert(TransactLineRepository::new(pool.clone()));
+    repositories.insert(UserAccountRepository::new(pool.clone()));
+
+    repositories
+}
 
 #[cfg(feature = "mock")]
-fn main() -> Result<(), &'static str> {
-    Err("Compiled with the mock feature enabled, server not supported in this mode")
+pub fn get_repositories() -> RepositoryMap {
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/main.rs:67]: @wlthomson

**Comment by wlthomson** [comment]
> 
> 
> Looks good, just a minor change request.
> 
> (haven't reviewed any mock stuff thought, would have been better in seperate PR imo). There is another way to do testing, which is fast and uses db, I'll make an issue with explanation. Anyways if we choose to go with mocks I'll give it a more thorough review in the future.
> 
> Also would be great if main method jumped up an abstraction level (still have logic in main.rs, but with helper methods), we can do this in seperate PR in the future, once we know what are all the bits that need to be in main
> 
> i.e. (something along these lines)
> 
> ```
> setupLogging();
> 
> let settings = getSetting();
> 
> let repository = getDBRepository(&settings)
> 
> startScheduler(&repository)
> 
> let syncConsumer = initiliseSync(&repository)
> 
> let startWebServer(&repository, &settings)
> ```

Yup, agreed 100%. Idea is that `main.rs` should just be an entry-point which uses functionality exposed by the crate. So I agree completely with lifting business logic out into functions etc.

File: src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/main.rs:67]: Agreed. I think this is an instance of your main comment about needing more abstraction in `main.rs`. Maybe best to make an issue so this stuff can get merged in the meantime?

**Review by andreievg — APPROVED** [review]

**Pull request merged**